File: backend/app/uploads.py
from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user
from app.db import get_session
from app.models import UploadedAsset, User
from app.schemas import (
    UploadCompleteRequest,
    UploadCompleteResponse,
    UploadPresignRequest,
    UploadPresignResponse,
)
from app.storage import ObjectStorageService
from app.config import settings


router = APIRouter(prefix="/api/uploads", tags=["uploads"])

MAX_UPLOAD_SIZE_BYTES = 20 * 1024 * 1024


def get_storage_service() -> ObjectStorageService:
    return ObjectStorageService(settings)


@router.post("/presign", response_model=UploadPresignResponse)
async def presign_upload(
    payload: UploadPresignRequest,
    user: Annotated[User, Depends(get_current_user)],
    storage: Annotated[ObjectStorageService, Depends(get_storage_service)],
) -> UploadPresignResponse:
    if payload.size_bytes > MAX_UPLOAD_SIZE_BYTES:
        raise HTTPException(status_code=413, detail="File exceeds 20MB limit")

    upload_id = uuid.uuid4()
    object_key = storage.build_upload_object_key(
        user_id=user.user_id,
        upload_id=upload_id,
        filename=payload.filename,
    )
    presigned = storage.build_presigned_upload_url(object_key)
    return UploadPresignResponse(
        upload_id=str(upload_id),
        object_key=object_key,
        upload_url=presigned.url,
        expires_in=presigned.expires_in,
    )


@router.post("/complete", response_model=UploadCompleteResponse)
async def complete_upload(
    payload: UploadCompleteRequest,
    user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_session)],
) -> UploadCompleteResponse:
    expected_prefix = f"uploads/{user.user_id}/"
    if not payload.object_key.startswith(expected_prefix):
        raise HTTPException(status_code=403, detail="Upload does not belong to user")

    asset = UploadedAsset(
        user_id=user.user_id,
        job_id=None,
        filename=payload.filename,
        mime_type=payload.mime_type,
        size_bytes=payload.size_bytes,
        object_key=payload.object_key,
        purpose=None,
    )
    db.add(asset)
    await db.flush()
    await db.commit()

    return UploadCompleteResponse(
        asset_id=str(asset.id),
        filename=asset.filename,
        mime_type=asset.mime_type,
        size_bytes=asset.size_bytes,
    )

File: frontend/src/api/uploads.ts
import { requestJson } from "./client";

export const MAX_CREATE_UPLOAD_SIZE_BYTES = 20 * 1024 * 1024;

export type UploadDraft = {
  upload_id: string;
  object_key: string;
  upload_url: string;
  expires_in: number;
};

export type UploadedAsset = {
  asset_id: string;
  filename: string;
  mime_type: string;
  size_bytes: number;
  object_key?: string;
};

export async function presignUpload(file: File): Promise<UploadDraft> {
  const response = await requestJson<UploadDraft>("/api/uploads/presign", {
    method: "POST",
    body: JSON.stringify({
      filename: file.name,
      mime_type: file.type || "application/octet-stream",
      size_bytes: file.size,
    }),
  });
  if (!response) {
    throw new Error("Upload presign returned an empty response.");
  }
  return response;
}

export async function uploadFileBinary(uploadUrl: string, file: File): Promise<void> {
  const response = await fetch(uploadUrl, {
    method: "PUT",
    body: file,
    headers: {
      "Content-Type": file.type || "application/octet-stream",
    },
  });
  if (!response.ok) {
    throw new Error("文件上传失败，请稍后重试。");
  }
}

export async function completeUpload(draft: UploadDraft, file: File): Promise<UploadedAsset> {
  const response = await requestJson<UploadedAsset>("/api/uploads/complete", {
    method: "POST",
    body: JSON.stringify({
      upload_id: draft.upload_id,
      object_key: draft.object_key,
      filename: file.name,
      mime_type: file.type || "application/octet-stream",
      size_bytes: file.size,
    }),
  });
  if (!response) {
    throw new Error("Upload complete returned an empty response.");
  }
  return {
    ...response,
    object_key: response.object_key ?? draft.object_key,
  };
}
